lib/jxl/enc_modular.cc

```diff
@@ -1421,6 +1421,13 @@ void ModularFrameEncoder::AddVarDCTDC(const Image3F& dc, size_t group_index,
   if (cparams.speed_tier >= SpeedTier::kSquirrel) {
     stream_options[stream_id].tree_kind = ModularOptions::TreeKind::kWPFixedDC;
   }
+  if (cparams.speed_tier < SpeedTier::kSquirrel && !nl_dc) {
```

Comment:
Is nl_dc false IFF we are doing JPEG recompression? I'd be slightly surprised...

Reply:
Oh you're right, it's false IF we do JPEG recompression but it can also be false in other cases (when distance < 2).

So the way I wrote it also makes regular VarDCT spend more time on DC at effort 8 and 9. That was not intended, but it's probably not a bad idea, actually. At those slow speeds, it's doing a lot of stuff anyway, so spending a bit more time on DC is not going to be a huge problem, I think.

Some quick results: (timing is not accurate at all, laptop was running hot after doing the 'before')

Before:
```
Encoding      kPixels    Bytes          BPP  E MP/s  D MP/s     Max norm        pnorm       BPP*pnorm   Bugs
------------------------------------------------------------------------------------------------------------
jxl:8           57482 11026557    1.5345979   0.304  19.021   2.80633187   0.62708369  0.962321348304      0
Aggregate:      57482 11026557    1.5345979   0.304  19.021   2.80633187   0.62708369  0.962321348304      0
```

After:

```
Encoding      kPixels    Bytes          BPP  E MP/s  D MP/s     Max norm        pnorm       BPP*pnorm   Bugs
------------------------------------------------------------------------------------------------------------
jxl:8           57482 11006224    1.5317681   0.266  15.616   2.80633187   0.62708369  0.960546825216      0
Aggregate:      57482 11006224    1.5317681   0.266  15.616   2.80633187   0.62708369  0.960546825216      0
```

So for e8 that's a 0.18% overall density improvement for d1, at the cost of slower encode and decode. Not sure if that's worth it, but it may be (after all, if you really care about encode/decode speed more than about compression, you'll likely not use -e 8 and --faster_decoding 0).

Reply:
the impact on vardct is likely much more at high distance where dc is more bytes (can be up to half at high distances like d32 range), i.e., this may help to catch up a little with AVIF at super low qualities

Reply:
Currently vardct uses three types of DC encoding:

- d<2: it just quantizes the DC and stores the quantized DC losslessly <-- this is where spending more effort on that lossless encoding does help for compression (though it hurts for enc/dec speed when it uses non-specialized code paths)
- 2 <= d < 4.5: it quantizes the DC in a way that is near-lossless: basically quantized predictor residuals -1, 0, 1 are kept as is (to have enough precision to avoid banding in slow gradients) and other residuals get their lsb set to zero.
- d > 4.5: a lossy modular DC frame is used, using quantization of Squeeze residuals

For the d > 4.5 case, we could look at ways to tune the DC frame encoding — currently it just does the same thing as if it was just any image to be encoded lossily with modular, which is likely not the best thing to do.

I'm personally not very interested in d > 8 because I think it has no practical applications whatsoever except pleasing people who like to evaluate image codecs at ridiculously low bitrates.